rust-runtime/smithy-json/src/deserialize/token.rs

```diff
@@ -0,0 +1,362 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+use crate::deserialize::error::{Error, ErrorReason};
+use crate::escape::unescape_string;
+use smithy_types::Number;
+use std::borrow::Cow;
+
+pub use crate::escape::Error as EscapeError;
+
+/// New-type around `&str` that indicates the string is an escaped JSON string.
+/// Provides functions for retrieving the string in either form.
+#[derive(Debug, PartialEq, Eq)]
+pub struct EscapedStr<'a>(&'a str);
+
+impl<'a> EscapedStr<'a> {
+    pub fn new(value: &'a str) -> EscapedStr<'a> {
+        EscapedStr(value)
+    }
+
+    /// Returns the escaped string value
+    pub fn as_escaped_str(&self) -> &str {
+        self.0
+    }
+
+    /// Unescapes the string and returns it.
+    /// If the string doesn't need unescaping, it will be returned directly.
+    pub fn to_unescaped(&self) -> Result<Cow<'a, str>, EscapeError> {
+        unescape_string(self.0)
+    }
+}
+
+/// Represents the location of a token
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+pub struct TokenOffset(pub usize);
```

Comment:
I would probably just name this `Offset`. We're already in a module named `Token`

Reply:
Makes sense.